### PR TITLE
Log worker info on error

### PR DESCRIPTION
**Pull request by j-bennet**
Different version of https://github.com/dask/distributed/pull/7847.

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

How this works:

```python
import distributed

class BadNumbers(Exception):
    def __str__(self):
        return "I didn't like these numbers: " + " ".join(f"{x:.1f}" for x in self.args)

def f():
    raise BadNumbers(1, 2.5, 3)

client = distributed.Client()
client.submit(f).result()

2023-07-20 15:02:50,741 - distributed.worker - WARNING - Compute Failed
Key:       f-b49acabd688e43ee290e83e090a617ea
Function:  f
args:      ()
kwargs:    {}
Exception: 'BadNumbers(1, 2.5, 3)'

2023-07-20 15:02:50,743 - distributed.client - ERROR - Compute Failed
Key:       f-b49acabd688e43ee290e83e090a617ea
Function:  f
args:      ()
kwargs:    {}
Exception: BadNumbers(1, 2.5, 3)
Worker:    {'tcp://127.0.0.1:63910'}
---------------------------------------------------------------------------
BadNumbers                                Traceback (most recent call last)
Cell In[3], line 1
----> 1 client.submit(f).result()

Cell In[2], line 8, in f()
      7 def f():
----> 8     raise BadNumbers(1, 2.5, 3)

BadNumbers: I didn't like these numbers: 1.0 2.5 3.0
```

### CODE REVIEW

File: distributed/client.py [hunk 1]
```diff
@@ -333,6 +333,9 @@ async def _result(self, raiseit=True):
         await self._state.wait()
         if self.status == "error":
             exc = clean_exception(self._state.exception, self._state.traceback)
+            logger.error(
+                f"in task: {self._state.key}\non worker:{self._state.erred_on}"
+            )
```
Comment (crusaderky) [inline distributed/client.py:338]: This will look really weird if anything captures the log. Like most production clients will.

**Comment by crusaderky** [comment]
Instead of:
```python
2023-06-23 15:37:47,237 - distributed.worker - WARNING - Compute Failed
Key:       f-78eea523b6ed747be234c3bfbc24c116
Function:  f
args:      ()
kwargs:    {}
Exception: 'BadNumbers(1, 2.5, 3)'

2023-06-23 15:37:47,238 - distributed.client - ERROR - in task: f-78eea523b6ed747be234c3bfbc24c116
on worker:{'tcp://127.0.0.1:54346'}
```

Wouldn't it be much cleaner and simpler to just add a line to the already existing message?
```python
2023-06-23 15:37:47,237 - distributed.worker - WARNING - Compute Failed
Key:       f-78eea523b6ed747be234c3bfbc24c116
Function:  f
args:      ()
kwargs:    {}
Exception: 'BadNumbers(1, 2.5, 3)'
Worker:    tcp://127.0.0.1:54346
```

**Comment by j-bennet** [comment]
@crusaderky

> Wouldn't it be much cleaner and simpler to just add a line to the already existing message?

It's not so easy, unfortunately. I wish every Python exception had something like a read-write `message` property that you could easily modify, and we would not need to jump through all these hoops, but that's not the case.

This PR is based on your suggestion here:

https://github.com/dask/distributed/pull/7847#issuecomment-1604235640

**Comment by j-bennet** [comment]
@crusaderky I made changes as we discussed. The message emitted on client is now very similar to the one emitted on worker.

File: distributed/client.py [hunk 2]
```diff
@@ -603,14 +616,47 @@ def set_error(self, exception, traceback):
...
 
+        self.key = key
+        self.erred_on = erred_on
         self.status = "error"
         self.exception = exception
         self.traceback = traceback
+        if run_spec:
+            func_args = run_spec.get("args")
+            func_kwargs = run_spec.get("kwargs")
+            self.func_args = pickle.loads(func_args) if func_args else ()
```
Comment (j-bennet) [inline distributed/client.py:636]: @crusaderky `pickle.loads` is probably not the right way to do this. What's the right way?

Comment (crusaderky) [reply distributed/client.py:636]: It's complicated. See
https://github.com/dask/distributed/blob/a7f7764ccf68a5efd10b6b3ee7227af49d7000a5/distributed/worker.py#L2206-L2222
and
https://github.com/dask/distributed/blob/a7f7764ccf68a5efd10b6b3ee7227af49d7000a5/distributed/worker.py#L2930-L2947

Please call `distributed.worker._deserialize` from here.

In rare use cases, run_spec may be `None` - for example if scattered data fails to deserialize. Please deal with the use case, but don't lose time writing a unit test for it - there are bugs related to it.

File: distributed/client.py [hunk 3]
```diff
@@ -603,14 +616,47 @@ def set_error(self, exception, traceback):
...
+        return textwrap.dedent(
+            f"""\
+Compute Failed
+Key:       {self.key}
+Function:  {func_name}
+args:      {self.func_args}
+kwargs:    {self.func_kwargs}
+Exception: {self.exception!r}
+Worker:    {self.erred_on}"""
+        )
```
Comment (crusaderky) [inline distributed/client.py:658]: Thanks to dedent you can align the text with the `f"""`

File: distributed/client.py [hunk 4]
```diff
@@ -1630,10 +1676,12 @@ def _handle_retried_key(self, key=None):
...
             state.retry()
 
-    def _handle_task_erred(self, key=None, exception=None, traceback=None):
+    def _handle_task_erred(
+        self, key=None, exception=None, traceback=None, erred_on=None, run_spec=None
+    ):
         state = self.futures.get(key)
         if state is not None:
-            state.set_error(exception, traceback)
+            state.set_error(exception, traceback, erred_on, key, run_spec)
```
Comment (crusaderky) [inline distributed/client.py:1684]: Consider:

```python
x = c.submit(lambda: 1/0, key="x")
y = c.submit(inc, x, key="y")
y.result()
```
I understand here that key="y"? However, the error happened while computing x, and x is what should appear in the error message.

File: distributed/client.py [hunk 5]
```diff
@@ -3257,6 +3306,9 @@ def get(
                     should_rejoin = False
             try:
                 results = self.gather(packed, asynchronous=asynchronous, direct=direct)
+            except Exception:
+                exc_type, exc_value, exc_traceback = sys.exc_info()
+                raise exc_value.with_traceback(exc_traceback)
```
Comment (crusaderky) [inline distributed/client.py:3311]: This makes no sense to me.
Are you saying that there was no traceback before?
Also, this code is specific to the sync use case; what happens on async clients?

File: distributed/queues.py [hunk 6]
```diff
@@ -267,7 +267,9 @@ def process(d):
             if d["type"] == "Future":
                 value = Future(d["value"], self.client, inform=True, state=d["state"])
                 if d["state"] == "erred":
-                    value._state.set_error(d["exception"], d["traceback"])
+                    value._state.set_error(
+                        d["exception"], d["traceback"], None, None, None
```
Comment (crusaderky) [inline distributed/queues.py:271]: This is hard to read; please use kwargs

File: distributed/tests/test_client.py [hunk 7]
```diff
@@ -401,18 +401,34 @@ async def test_future_tuple_repr(c, s, a, b):
...
+    with captured_logger("distributed.client") as logger:
+        x = c.submit(div, 1, 0)
+        assert isinstance(x.exception(), ZeroDivisionError)
+    assert "Worker: " in logger.getvalue()
+
+    with captured_logger("distributed.client") as logger:
+        x = c.submit(div, 1, 0)
+        with pytest.raises(ZeroDivisionError):
+            _ = x.result()
+    assert "Worker: " in logger.getvalue()
```
Comment (crusaderky) [inline distributed/tests/test_client.py:431]: please add test for x = 1/0; y = x + 1; y.result()

File: distributed/variable.py [hunk 8]
```diff
@@ -208,7 +208,7 @@ async def _get(self, timeout=None):
         if d["type"] == "Future":
             value = Future(d["value"], self.client, inform=True, state=d["state"])
             if d["state"] == "erred":
-                value._state.set_error(d["exception"], d["traceback"])
+                value._state.set_error(d["exception"], d["traceback"], None, None, None)
```
Comment (crusaderky) [inline distributed/variable.py:211]: please use kwargs

**Review by crusaderky — CHANGES_REQUESTED** [review]
.

File: distributed/client.py (hunk 5 above)
Comment (j-bennet) [reply distributed/client.py:3311]: Good catch. I think this is a leftover from the first version of this PR, where I'd only add worker info to exception if any `client` frames were present in traceback; completely unnecessary here.

File: distributed/client.py (hunk 4 above)
Comment (j-bennet) [reply distributed/client.py:1684]: Yes, but how would you know that here? go through task dependencies? what if more than one failed?

Comment (j-bennet) [reply distributed/client.py:1684]: Ah. I see there's `exception_blame`.

**Pull request closed without merging**